### DISCUSSION TITLE
noexcept or NoExceptionAttribute for async methods

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#6079] [Language Ideas]
## Background
In the current version of C#, an async method will yield a state machine with `try-catch` blocks in the `MoveNext` method. But this would prevent the JIT optimizing (especially, inlining) the `MoveNext`, which will greatly harm the performance of async methods.

## Proposal
Introduce a keyword `noexcept` or an attribute `NoExceptionAttribute` for async methods, then compiler won't generate try-catch blocks for methods annotated with `noexcept` or `[NoException]`.

This could also be extended to other scenarios to suppress compilers emitting `try-catch` in auto-generated code.

## Usage
```csharp
noexcept async Task<int>(int x)
{
    await Bar();
    return x;
}
```
or

```csharp
[NoException]
async Task<int>(int x)
{
    await Bar();
    return x;
}
```

and the compiler will lower it to
```csharp
[CompilerGenerated]
struct FooAsyncStateMachine
{
    ...
    public void MoveNext()
    {
        // no try-catch blocks emitted
    }
}
```

This is especially useful for `ValueTask<T>` scenarios:

```csharp
Console.WriteLine(await Foo());

noexcept ValueTask<int> Foo()
{
    return ValueTask.FromResult(1);
}
```

Now we will have the potential to inline the async method call.

**Comment by jnm2** [comment]
What happens when an exception is thrown then?

**Reply by hez2010** [reply]
`noexcept` is guaranteed by users, so in this case the exception will be omitted away, which is the same behavior with `async void`.

**Reply by HaloFour** [reply]
That behavior is an anti-pattern which is why `async void` should be avoided unless absolutely required.

**Reply by hez2010** [reply]
`noexcept` is only useful in high performance async code scenarios, when users use `noexcept`, they should be responsible to what they are authoring.  
And we can definitely put the `noexcept` under unsafe.

**Reply by HaloFour** [reply]
IMO that's an oxymoron, the overhead of the `async` method itself is absolutely nothing compared to the overhead of any async code that will actually execute.

Swallowing an exception is also the wrong behavior for `noexcept`, in other languages that `try`/`catch` is still emitted but it asserts rather than throwing.  The user code doesn't see the exception because the user code is terminated.  Swallowing the exception would certainly be the wrong behavior here.

**Reply by hez2010** [reply]
> the overhead of the `async` method itself is absolutely nothing compared to the overhead of any async code that will actually execute

Imagine you have an async method:
```csharp
var x = await Foo();
Console.WriteLine(x);
```

while the method `Foo` is:
```csharp
ValueTask<int> Foo()
{
    return ValueTask.FromResult(1);
}
```

In this case, the `try-catch` becomes the major overhead of the async method, even we already use `ValueTask`.

With `noexcept` we will have the potential to inline the async method call.

**Reply by HaloFour** [reply]
I'd suggest that the compiler (or runtime) could consider eliding the `try`/`catch`, but I think it would be completely wrong for the compiler to add special syntax to do something so fundamentally wrong like swallowing exceptions for the incredibly niche case of someone writing async methods that don't do anything.  The use of `ValueTask<T>` is already a can of worms.

**Reply by HaloFour** [reply]
This is also the opposite of what I'd expect `noexcept` to do.  In C++, when a function is marked as `noexcept` it still emits a `try`/`catch` to ensure that exceptions cannot bubble out, and any exception actually thrown is caught and results in a call to `std::terminate`.

**Reply by jnm2** [reply]
> `noexcept` is guaranteed by users, so in this case the exception will be omitted away, which is the same behavior with `async void`.

So then you are still emitting try/catch? If not, then the exception bubbles up, just not to a sensible place.

**Reply by hez2010** [reply]
> > `noexcept` is guaranteed by users, so in this case the exception will be omitted away, which is the same behavior with `async void`.
> 
> So then you are still emitting try/catch? If not, then the exception bubbles up, just not to a sensible place.

My bad. It's not the same with `async void`, instead it throws an unhandled exception.

**Reply by HaloFour** [reply]
You can do this today by removing the `async` modifier from the method.

**Reply by jnm2** [reply]
@hez2010 I don't think this is safe in any way. The exception bubbles up to a place that is hard to reason about. It might forcibly terminate the application (if it's a bare thread) or get picked up in a message loop exception handler or thread pool exception handler.

**Reply by jnm2** [reply]
For all I know, it might leave a state machine somewhere in the async call stack in an invalid state from which the application tries to recover with bad results.

**Comment by alexrp** [comment]
`noexcept` would imply no allocations. What `async` code can actually function with zero allocations under any circumstances?

**Reply by huoyaoyuan** [reply]
Are OOM/SO treated as unrecoverable?

**Comment by TahirAhmadov** [comment]
To my knowledge, the exception wouldn't even be swallowed - it would bubble up that worker thread's call stack and crash your application.
Also, I'm not convinced `MoveNext` being inlined is prevented by the `try-catch`; if I'm not mistaken, they use the generic constraint interface trick to avoid boxing the state machine, and `MoveNext` becomes essentially an `Opcodes.Call`, and it's easy to inline it already.
Don't quote me on the points above, but I'm >70% I'm correct on both.

**Reply by hez2010** [reply]
> being inlined is prevented by the try-catch

The JIT will refuse to inline any method contains `try-catch` blocks.

**Comment by Neme12** [comment]
I don't think you can really guarantee that exceptions won't be thrown in any meaningful piece of code. The runtime can throw exceptions almost everywhere.

**Comment by theunrepentantgeek** [comment]
> In the current version of C#, an async method will yield a state machine with try-catch blocks in the MoveNext method. But this would prevent the JIT optimizing (especially, inlining) the MoveNext

This is less an argument that the C# language needs to change and more one that there's room for improvement in the JIT compiler - if it could be enhanced to inline methods as you suggest, the benefit would apply across a far wider range of methods and runtimes than a narrowly scoped niche language change.

> which will greatly harm the performance of async methods.

Have you benchmarked the overhead of the async state machine? 

I haven't done so myself, but my intuition is that the overhead would be on the close order of 10s or 100s of nanoseconds, which pales into insignificance against the thread scheduling overhead of the OS (usually milliseconds) or round-trip networking (typically milliseconds, but sometimes microseconds in specialist cases).
